metron-platform/metron-common/src/main/java/org/apache/metron/common/configuration/manager/ZkConfigurationManager.java

```diff
@@ -0,0 +1,129 @@
+ *  limitations under the License.
+ *
+ */
+
+package org.apache.metron.common.configuration.manager;
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.cache.NodeCache;
+import org.apache.curator.utils.CloseableUtils;
+import org.apache.metron.common.utils.JSONUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.apache.commons.lang.ArrayUtils.isNotEmpty;
+
+/**
+ * Responsible for managing configuration values that are created, persisted, and updated
+ * in Zookeeper.
+ */
+public class ZkConfigurationManager implements ConfigurationManager {
+
+  /**
+   * A Zookeeper client.
+   */
+  private CuratorFramework zookeeperClient;
+
+  /**
+   * The configuration values under management.  Maps the path to the configuration values
+   * in Zookeeper to the cache of its values.
+   */
+  private Map<String, NodeCache> valuesCache;
+
+  /**
+   * @param zookeeperClient The client used to communicate with Zookeeper.  The client is not
+   *                        closed.  It must be managed externally.
+   */
+  public ZkConfigurationManager(CuratorFramework zookeeperClient) {
+    this.zookeeperClient = zookeeperClient;
+    this.valuesCache = Collections.synchronizedMap(new HashMap<>());
+  }
+
+  /**
+   * Define the paths within Zookeeper that contains configuration values that need managed.
+   * @param zookeeperPath The Zookeeper path.
+   */
+  @Override
+  public ZkConfigurationManager with(String zookeeperPath) {
+    NodeCache cache = new NodeCache(zookeeperClient, zookeeperPath);
+    valuesCache.put(zookeeperPath, cache);
+    return this;
+  }
+
+  /**
+   * Open a connection to Zookeeper and retrieve the initial configuration value.
+   */
+  @Override
+  public synchronized ZkConfigurationManager open() throws IOException {
+    try {
+      doOpen();
+    } catch(Exception e) {
+      throw new IOException(e);
+    }
+
+    return this;
+  }
+
+  private void doOpen() throws Exception {
+    for (NodeCache cache : valuesCache.values()) {
+      cache.start(true);
+    }
+  }
+
+  /**
+   * Retrieve the configuration object.
+   */
+  @Override
+  public synchronized <T> Optional<T> get(String key, Class<T> clazz) throws IOException {
+    T result = null;
+
+    NodeCache cache = valuesCache.get(key);
+    if(cache != null && cache.getCurrentData() != null && isNotEmpty(cache.getCurrentData().getData())) {
+      result = deserialize(cache.getCurrentData().getData(), clazz);
+    }
+
+    return Optional.ofNullable(result);
+  }
+
+  /**
+   * Close the configuration manager.
+   *
+   * Does not close the zookeeperClient that was passed in to the constructor.
+   */
+  @Override
+  public synchronized void close() {
+    for (NodeCache cache : valuesCache.values()) {
+      CloseableUtils.closeQuietly(cache);
```

Comment:
The ZkConfigurationManager does not open its own client.  The client is passed into the constructor.  I expect the client to be managed externally.  Its mentioned in the constructor javadoc.

Also, I thought it appropriate per the usage guideline in the [Curator docs](http://curator.apache.org/curator-framework/index.html)...

> IMPORTANT: CuratorFramework instances are fully thread-safe. You should share one CuratorFramework per ZooKeeper cluster in your application.